### ISSUE TITLE
[scatter] Axes don't match array dimensions in RepetitionProcessor with non-trivial penalty

### ISSUE THREAD

**Issue by antmikinka** [lemonade-sdk/lemon-mlx-engine#32]
## Bug

When `repetition_penalty != 1.0` is configured, text generation fails with:

```
Generation error: [scatter] Axes don't match array dimensions.
```

## Root Cause

`src/common/generate.cpp:188` passes `-1` as the axis argument to `mx::scatter()`:

```cpp
auto result = mx::scatter(logits, shaped_indices, penalized, -1);
```

Unlike other MLX ops (e.g. `take_along_axis`, `concatenate`, `stack`), **MLX C++ `scatter()` does not normalize negative axis values**. The host-side validation in `mlx/ops.cpp:3596` rejects negative axes:

```cpp
if (*dims.begin() < 0 || *dims.rbegin() >= a.ndim()) {
    throw std::invalid_argument("[scatter] Axes don't match array dimensions.");
}
```

This is also a limitation in upstream MLX — `scatter()` and `gather()` are inconsistent with other axis-based ops that call `normalize_axis_index()`.

## Reproduction

Configure any model with `repetition_penalty != 1.0` and generate text. The error fires as soon as the `RepetitionProcessor::process()` method executes the scatter path.

## Fix

Replace `-1` with `logits.ndim() - 1`:

```cpp
auto result = mx::scatter(logits, shaped_indices, penalized, logits.ndim() - 1);
```

- For 2D logits `[B, V]`: resolves to `1` (last axis)
- For 1D logits `[V]`: resolves to `0` (last axis)

## Impact

- Pre-existing bug — never caught because existing unit tests use `penalty=1.0` which hits an early return
- Affects all backends (CPU, ROCm, Metal) — scatter validation is host-side
- Unrelated to the ROCm `clear_streams()` patch

**Comment by antmikinka** [comment]
#31  IS from an old branch. cleaning up by new branch, means new PR. Trying to keep it 1 branch : 1 issue : 1 pr.
Therefore that is #33
